# Import needed libraries

In [1]:
# python library to perform numeric operations on numpy arrays
import numpy as np

# python library to process dataframe and do analysis on them
import pandas as pd

# !pip install xlrd
# python library to deal with excel files
import xlrd

# library to deal with directories effectively
import glob 

# convert categorical feature into numbers (specifically the ownerhip status)
from sklearn import preprocessing


import matplotlib.pyplot as plt
import seaborn as sns

## Read in and merge startup data files Christoph gave

In [2]:
# Using the Pandas python library
# Load the excel datafile
# Located in the data directory which is one directory above hence the ..


path = '../data/' # use your path

allFiles = glob.glob(path + "/*.xlsx") #  all excel files


frame = pd.DataFrame()
final = []

#  merge all excel files into one
for datafiles in allFiles:
    df = pd.read_excel(datafiles, skiprows=7, skip_footer=3)
    final.append(df)
    
pitchbk_data = pd.concat(final)

#  reset indices
pitchbk_data = pitchbk_data.reset_index(drop=True)

## Investigate merged dataset

In [3]:
# Display a verbose description of the dataset
# This includes number of columns, datatype of each column, how much memory the file takes up

pitchbk_data.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1655 entries, 0 to 1654
Data columns (total 130 columns):
Company ID                             object
Company Name                           object
Company Former Name                    object
Company Also Known As                  object
PBId                                   object
Description                            object
Primary Industry Sector                object
Primary Industry Group                 object
Primary Industry Code                  object
All Industries                         object
Industry Vertical                      object
Company Financing Status               object
Total Raised                           float64
Business Status                        object
Ownership Status                       object
Universe                               object
Website                                object
Employees                              float64
Exchange                               object
Ticker          

In [4]:
pitchbk_data

Company ID                   Company Name Company Former Name  \
0      61943-41                 1001Pharmacies                 NaN   
1     126697-60                      24Storage                 NaN   
2      60359-86                           2Can                 NaN   
3      81674-02                    360imprimir                 NaN   
4      58634-74                        3D Hubs                 NaN   
5     172624-96                   A Cloud Guru                 NaN   
6     108577-54                          Acast       Acast Stories   
7      64576-81                      AccessPay                 NaN   
8     167047-57          Achilles Therapeutics          AchillesTx   
9      58006-72                        Adbrain                 NaN   
10     54120-79               ADC Therapeutics                 NaN   
11     54540-10                         adjust              Adeven   
12     56199-34                         Admedo              Adizio   
13    109569-34                        AdotMob                 NaN   
14     56856-52                     advanceCOR                 NaN   
15    151170-13             Aelix Therapeutics                 NaN   
16     62510-14                     AfriMarket                 NaN   
17    154990-45                       Agricool                 NaN   
18    126199-27                         Aifloo                 NaN   
19    120982-15                       AImotive           AdasWorks   
20     81891-91                        Aircall                 NaN   
21    111733-48                      Aire Labs                 NaN   
22     62510-41                         Akeneo                 NaN   
23    167531-14        Alan (health insurance)                 NaN   
24     59252-32                        Algolia                 NaN   
25     61923-16                         Algomi                 NaN   
26     56648-08           Allecra Therapeutics                 NaN   
27    158293-09       Alpha Ring International                 NaN   
28     62440-93                      AltraTech                 NaN   
29     61586-92              Amal Therapeutics                 NaN   
...         ...                            ...                 ...   
1625  121608-73                    Push Doctor                 NaN   
1626   64829-53                   Book a Tiger                 NaN   
1627   98734-24                       Callsign                 NaN   
1628  169825-87          Gilo Industries Group                 NaN   
1629   94417-57                        Unlockd                 NaN   
1630   59226-04                  simplesurance                 NaN   
1631  118093-51                   Ultrahaptics                 NaN   
1632  156548-17                        Tricoya                 NaN   
1633  168804-55                     Bink (App)                 NaN   
1634  125143-57               Scalable Capital                 NaN   
1635  161211-07                       Ottonova                 NaN   
1636  108317-08                        Arralis                 NaN   
1637  100059-04                       Helpling                 NaN   
1638  110556-19                Cantab Research                 NaN   
1639  157761-28             CompareEuropeGroup                 NaN   
1640  113444-02                 Babylon Health                 NaN   
1641  104383-72                        Revolut                 NaN   
1642  166649-68                      Graphcore                 NaN   
1643  179124-31     Picnic (platform software)                 NaN   
1644   57321-01                Green Biologics                 NaN   
1645  118978-21                          Monzo               Mondo   
1646  152084-89                         Neyber                 NaN   
1647   99412-84                 Future Finance                 NaN   
1648   62352-64               Landbay Partners                 NaN   
1649   91141-75                     Improbable           Macsco 47   
1650  175366-27

Here, the Pitchbook data says that there are 1655 rows (entries), indexed from 0 to 1654.
In addition, there are 130 columns of different datatypes (many objects like strings (ie phrases or names), a good number of numeric values (floats) and a couple date and time formatted columns). 

The result of this information implies that some cleaning and encoding of the dataset will need to be done to present it in an effective structure for the Machine Learning algorithms later on.

Finally, the dataset takes up to ~1.5 MB (a thousandth of a Gigabyte).

In [5]:
# Display the dimension of the dataframe 
# which from the info above should be 1655 rows and 130 columns

pitchbk_data.shape

(1655, 130)

(1655,130) indicates that there are 1655 startups in the dataframe and 130 attributes/features about each startup.

In [6]:
# encode names correctly
list_col_names = [name.encode('utf-8') for name in pitchbk_data.columns.values]

# get first 20 features
print(list_col_names[:20])

['Company ID', 'Company Name', 'Company Former Name', 'Company Also Known As', 'PBId', 'Description', 'Primary Industry Sector', 'Primary Industry Group', 'Primary Industry Code', 'All Industries', 'Industry Vertical', 'Company Financing Status', 'Total Raised', 'Business Status', 'Ownership Status', 'Universe', 'Website', 'Employees', 'Exchange', 'Ticker']


In [7]:
pitchbk_data['Business Status'].unique()

array([u'Profitable', u'Generating Revenue', u'Startup',
       u'Clinical Trials - Phase 1', u'Product Development',
       u'Pre-Clinical Trials', u'Generating Revenue/Not Profitable',
       u'Product In Beta Test', u'Stealth', u'Clinical Trials - General',
       u'Clinical Trials - Phase 2', u'Bankruptcy: Admin/Reorg',
       u'Out of Business'], dtype=object)

In [8]:
pitchbk_data['Company Financing Status'].unique()

array([u'Venture Capital-Backed', u'Private Equity-Backed',
       u'Formerly VC-Backed', u'Angel-Backed',
       u'Accelerator/Incubator Backed', u'Corporate Backed or Acquired'], dtype=object)

In [9]:
pitchbk_data['Financing Status Note'] # news information about financial status

0       The company raised EUR 8 million of Series A v...
1       The company raised SEK 250 million of funding ...
2       The company raised $3 million of venture fundi...
3       The company raised EUR 4 million of Series B v...
4       The company raised $6.99 million of Series B v...
5       The company raised $7 million of Series A fund...
6       The company received an EUR 2.22 million grant...
7       The company received GBP 2 million of debt fin...
8       The company raised GBP 13.2 million of Series ...
9       The company raised $2 million of venture fundi...
10      The company raised $105 million of venture fun...
11      The company raised $17 million of venture fund...
12      The company raised $6 million of Series B vent...
13      Iris Capital Management sold its stake in the ...
14      The company received EUR 900,000 in grant fund...
15      The company raised $13 million of Series A ven...
16      The company raised EUR 10 million of venture f...
17      The co

In [10]:
pitchbk_data['Last Financing Deal Class'].unique()

array([u'Venture Capital', u'Other', u'Debt', u'Corporate', u'Individual',
       u'Private Equity', u'Out of Business', u'Public Investment',
       u'Bankruptcy'], dtype=object)

In [11]:
pitchbk_data['Ownership Status'].unique()

array([u'Privately Held (backing)',
       u'Acquired/Merged (Operating Subsidiary)', u'Acquired/Merged',
       u'Out of Business', u'Publicly Held'], dtype=object)

In [12]:
pitchbk_data['Ownership Status']

0                     Privately Held (backing)
1                     Privately Held (backing)
2                     Privately Held (backing)
3                     Privately Held (backing)
4                     Privately Held (backing)
5                     Privately Held (backing)
6                     Privately Held (backing)
7                     Privately Held (backing)
8                     Privately Held (backing)
9                     Privately Held (backing)
10                    Privately Held (backing)
11                    Privately Held (backing)
12                    Privately Held (backing)
13                    Privately Held (backing)
14                    Privately Held (backing)
15                    Privately Held (backing)
16                    Privately Held (backing)
17                    Privately Held (backing)
18                    Privately Held (backing)
19                    Privately Held (backing)
20                    Privately Held (backing)
21           

Ownership status appears to be a good label to do predictions on.

Before that though, I want to investigate the natural groupings/clusters present in the data and see what they might signify

## Pre-process/clean dataset

In [13]:
pitchbk_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1655 entries, 0 to 1654
Columns: 130 entries, Company ID to PitchBook Link
dtypes: datetime64[ns](2), float64(37), object(91)
memory usage: 1.6+ MB


In [14]:
# first and last financing date are the datetime attributes
pitchbk_data.select_dtypes(['datetime64'])

First Financing Date Last Financing Date
0              2013-03-03          2015-07-09
1              2015-05-01          2017-02-20
2              2012-10-08          2017-07-25
3              2014-04-30          2017-04-24
4              2013-04-01          2016-07-13
5              2017-07-31          2017-07-31
6              2014-02-10          2017-07-03
7              2012-07-26          2017-04-13
8              2016-10-05          2016-10-05
9              2013-07-17          2017-05-22
10             2012-03-26          2016-10-18
11             2012-04-01          2015-10-21
12             2013-02-22          2016-02-04
13             2015-03-02                 NaT
14             2013-03-01          2016-05-01
15             2016-01-11          2016-01-11
16             2013-07-10          2016-09-06
17             2015-05-11          2017-07-11
18             2015-06-01          2017-09-05
19                    NaT          2016-03-29
20             2014-01-09          2016-09-14
21             2014-06-16          2017-07-23
22             2013-09-30          2017-03-20
23             2016-10-24          2016-10-24
24             2014-01-01          2017-06-08
25             2012-12-01          2017-03-03
26             2013-04-18          2016-05-20
27             2016-04-29          2016-04-29
28             2014-01-08          2017-07-03
29                    NaT          2017-09-05
...                   ...                 ...
1625           2014-11-10          2017-07-05
1626           2014-07-10          2017-11-16
1627                  NaT          2017-07-27
1628           2012-11-29          2017-01-16
1629           2015-01-01          2017-05-24
1630           2012-05-01                 NaT
1631           2014-11-13          2017-05-03
1632           2012-10-08          2017-03-29
1633           2016-06-02          2017-03-18
1634           2015-01-19          2017-06-20
1635                  NaT          2017-07-04
1636           2014-12-18          2017-03-30
1637                  NaT          2017-08-01
1638                  NaT          2017-01-06
1639           2015-03-11          2017-01-11
1640                  NaT          2017-04-25
1641                  NaT          2017-07-30
1642                  NaT          2017-11-12
1643           2017-03-28          2017-03-28
1644           2005-12-22          2017-03-07
1645           2015-06-30          2017-11-07
1646           2015-01-01          2017-09-12
1647           2015-03-12          2017-07-28
1648           2013-09-24          2017-08-17
1649                  NaT          2017-07-20
1650           2017-08-23          2017-08-23
1651           2017-04-11          2017-04-11
1652           2017-07-17          2017-07-17
1653           2017-02-01          2017-02-01
1654           2017-03-01          2017-03-01

[1655 rows x 2 columns]

In [15]:
# full procedure to convert objects to categorical for easy processing
pitchbk_data = pd.concat([
        pitchbk_data.select_dtypes([], ['object', 'datetime64']),
        pitchbk_data.select_dtypes(['object', 'datetime64']).apply(pd.Series.astype, dtype='category')
        ], axis=1).reindex_axis(pitchbk_data.columns, axis=1)


In [16]:
# get all non numeric attributes
category_features = pitchbk_data.select_dtypes(['category']).columns

In [17]:
label_name = pitchbk_data["Ownership Status"]
label_num = pitchbk_data["Ownership Status"].cat.codes

In [18]:
label_name

0                     Privately Held (backing)
1                     Privately Held (backing)
2                     Privately Held (backing)
3                     Privately Held (backing)
4                     Privately Held (backing)
5                     Privately Held (backing)
6                     Privately Held (backing)
7                     Privately Held (backing)
8                     Privately Held (backing)
9                     Privately Held (backing)
10                    Privately Held (backing)
11                    Privately Held (backing)
12                    Privately Held (backing)
13                    Privately Held (backing)
14                    Privately Held (backing)
15                    Privately Held (backing)
16                    Privately Held (backing)
17                    Privately Held (backing)
18                    Privately Held (backing)
19                    Privately Held (backing)
20                    Privately Held (backing)
21           

In [19]:
label_num

0       3
1       3
2       3
3       3
4       3
5       3
6       3
7       3
8       3
9       3
10      3
11      3
12      3
13      3
14      3
15      3
16      3
17      3
18      3
19      3
20      3
21      3
22      3
23      3
24      3
25      3
26      3
27      3
28      3
29      3
       ..
1625    3
1626    3
1627    3
1628    3
1629    3
1630    3
1631    3
1632    1
1633    3
1634    3
1635    3
1636    3
1637    3
1638    3
1639    3
1640    3
1641    3
1642    3
1643    3
1644    3
1645    3
1646    3
1647    3
1648    3
1649    3
1650    3
1651    3
1652    3
1653    3
1654    3
Length: 1655, dtype: int8

In [20]:
# encode categorical features
for feature in category_features:
    pitchbk_data[feature] = pitchbk_data[feature].cat.codes

In [21]:
np.sort(pitchbk_data['Company ID'].unique())

array([   0,    1,    2, ..., 1453, 1454, 1455])

1456 unique startups present out of the 1655 startups, showing that some startups may have some timescale of information added to the dataset, this could help understand how startup features change from one time point to another

In [22]:
#axis=1 relates to columns
# if all values in a column are NaN, the column is dropped
clean_numpitchbk = pitchbk_data.dropna(axis=1, how='all')

In [23]:
pitchbk_data.columns[pitchbk_data.isnull().all()].tolist()

[u'First Financing Deal Type 3',
 u'First Financing Debt Type 3',
 u'Last Financing Deal Type 3',
 u'Last Financing Debt Type 3',
 u'PitchBook Link']

These 5 colums are dropped as they contain null values throughout their columns

In [24]:
clean_numpitchbk

Company ID  Company Name  Company Former Name  Company Also Known As  \
0           1195             1                   -1                     -1   
1            302             4                   -1                     -1   
2           1159             5                   -1                     -1   
3           1285             8                   -1                     -1   
4           1122             9                   -1                     -1   
5            699            18                   -1                    210   
6            122            25                    1                     -1   
7           1247            26                   -1                     -1   
8            540            27                    2                     -1   
9           1107            35                   -1                     -1   
10           996            20                   -1                     -1   
11          1001          1426                    5                     -1   
12          1045            38                    6                     -1   
13           138            39                   -1                     -1   
14          1066          1427                   -1                     -1   
15           365            45                   -1                     -1   
16          1214            47                   -1                     -1   
17           406            48                   -1                     -1   
18           299            54                   -1                     -1   
19           259            21                    4                     -1   
20          1286            56                   -1                     -1   
21           159            57                   -1                      4   
22          1215            59                   -1                     -1   
23           555            60                   -1                     -1   
24          1137            63                   -1                     -1   
25          1192            64                   -1                     -1   
26          1063            65                   -1                     -1   
27           446            67                   -1                    159   
28          1213            69                   -1                     -1   
29          1188            70                   -1                     -1   
...          ...           ...                  ...                    ...   
1625         267           986                   -1                     -1   
1626        1257           198                   -1                     26   
1627        1422           233                   -1                     -1   
1628         621           542                   -1                     -1   
1629        1374          1310                   -1                     -1   
1630        1134          1456                   -1                     -1   
1631         221          1303                   -1                     -1   
1632         429          1284                   -1                     -1   
1633         590           163                   -1                     -1   
1634         289          1055                   -1                     -1   
1635         470           897                   -1                     -1   
1636         115            98                   -1                     -1   
1637           0           584                   -1                     -1   
1638         148           238                   -1                    174   
1639         438           300                   -1                     -1   
1640         172           139                   -1                     21   
1641          70          1026                   -1                     -1   
1642         527           554                   -1                     -1   
1643         762           946                   -1                    149   
1644        1087           5

3 rows were dropped (503-500)

In [25]:
for idx, elem in enumerate((clean_numpitchbk == clean_numpitchbk.iloc[0]).all()):
    if elem == True:
        print clean_numpitchbk.columns[idx]

HQ Global Region
Profile Data Source


In [26]:
# remove columns with just a constant value
clean_numpitchbk = clean_numpitchbk.loc[:, (clean_numpitchbk == clean_numpitchbk.iloc[0]).any()] 

2 columns were deleted as they had constant features: HQ Global Region and Profile Data Source. This is because all startups are within the european region. In addition, the Profile Data Source is just filled with the text "PitchBook Research"

Due to the class imbalance I am not going to use accuracy as the way to evaluate the performance of the classifier I end up using (for now)

For now Im going to set all NaN to the mean of the values in a feature/column

In [27]:
clean_numpitchbk = clean_numpitchbk.fillna(clean_numpitchbk.mean())

## Normalising data (by making features have a similar mean and variance) - store for now to see if performance improves later

In [28]:
df = clean_numpitchbk

# subtract each attribute by its mean
# divide by its standard deviation
normalised_numpitchbk = (df-df.mean()) / (df.std())

This dataset currently has 1655 samples and 125 features

## Investigate different clustering and dimensionality reduction techniques to visualise and understand data

In [29]:
from __future__ import print_function, division
%matplotlib inline

import collections


from sklearn.cluster import KMeans
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB

from sklearn.metrics import adjusted_rand_score, confusion_matrix
from sklearn.decomposition import PCA
from sklearn.model_selection import cross_val_score, KFold

In [30]:
clean_numpitchbk.describe()

Company ID  Company Name  Company Former Name  Company Also Known As  \
count  1655.000000   1655.000000          1655.000000            1655.000000   
mean    721.677946    725.709970             7.841088              15.240483   
std     423.861838    420.420618            29.335911              46.151309   
min       0.000000      0.000000            -1.000000              -1.000000   
25%     355.500000    361.500000            -1.000000              -1.000000   
50%     712.000000    727.000000            -1.000000              -1.000000   
75%    1098.500000   1089.000000            -1.000000              -1.000000   
max    1455.000000   1460.000000           159.000000             217.000000   

              PBId  Description  Primary Industry Sector  \
count  1655.000000  1655.000000              1655.000000   
mean    721.677946   734.694864                 3.511178   
std     423.861838   429.204322                 1.956841   
min       0.000000     0.000000                 0.000000   
25%     355.500000   363.500000                 1.000000   
50%     712.000000   733.000000                 5.000000   
75%    1098.500000  1109.500000                 5.000000   
max    1455.000000  1479.000000                 6.000000   

       Primary Industry Group  Primary Industry Code  All Industries  \
count             1655.000000            1655.000000     1655.000000   
mean                24.694260              53.645317      387.701511   
std                 11.046874              40.669271      246.868463   
min                  0.000000               0.000000        0.000000   
25%                 16.000000              19.000000      173.000000   
50%                 33.000000              49.000000      378.000000   
75%                 33.000000              86.500000      594.500000   
max                 35.000000             138.000000      861.000000   

              ...           Facebook Likes  Facebook Likes Change  \
count         ...              1655.000000            1655.000000   
mean          ...               418.330514              49.889709   
std           ...               368.532716             117.999268   
min           ...                -1.000000            -802.000000   
25%           ...                -1.000000               1.000000   
50%           ...               374.000000              12.000000   
75%           ...               737.500000              49.889709   
max           ...              1116.000000             979.000000   

       Facebook Likes % Change  Majestic Referring Domains  \
count              1655.000000                 1655.000000   
mean                 75.862236                  142.755996   
std                  90.718650                  176.301107   
min                  -1.000000                    1.000000   
25%                  -1.000000                   25.000000   
50%                  40.000000                   94.000000   
75%                 106.000000                  146.500000   
max                 365.000000                  986.000000   

       Majestic Referring Domains Change  Majestic Referring Domains % Change  \
count                        1655.000000                          1655.000000   
mean                            1.413844                           204.073716   
std                            25.027571                           143.628895   
min                          -342.000000                            -1.000000   
25%                            -1.000000                            60.500000   
50%                             0.000000                           266.000000   
75%                             1.413844                           266.000000   
max                           808.000000                           513.000000   

       Twitter Followers  Twitter Followers Change  \
count        1655.000000               1655.000000   
mean          482.352870                 12.462085   
std           359.187463

In [31]:
X = clean_numpitchbk.drop(labels=["Ownership Status"], axis=1, inplace=False)

# y is a vector of categories
y = np.array(label_name)

In [32]:
y = pd.DataFrame(y, columns=["Ownership Status"])

In [34]:
y['Ownership Status'].value_counts()

Privately Held (backing)                  1639
Acquired/Merged (Operating Subsidiary)       7
Acquired/Merged                              4
Publicly Held                                3
Out of Business                              2
Name: Ownership Status, dtype: int64